### PR TITLE
Tdd tests

### PR DESCRIPTION
**Pull request by bertil291utn**
### Testing TDD with RSpec in enumerable methods
We tested the following methods
- My_each
- My_each_with_index
- My_select
- My_all?
- My_any?
- My_none?
- My_count
- My_map
- My_inject

It was tested with different kind of variables (procs, range, arrays) and different situations

We really appreciate your feedback, thank you in beforehand 💻 🚀

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+      result = []
+      array_numbers.my_each { |elem| result.push(elem) }
+      expect(array_numbers).to eql(result)
+    end
+
+    it 'is returning same array as original string one' do
+      result = []
+      array_string.my_each { |elem| result.push(elem) }
+      expect(array_string).to eql(result)
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: [Optional]
As it is testing the same case, it returns the same array or not(Does not depend on the type of value it contains). You can combine it withing one test case. 
```ruby
it 'is returning same array as original one' do
      result = []
      array_numbers.my_each { |elem| result.push(elem) }
      expect(array_numbers).to eql(result)
      result = []
      array_string.my_each { |elem| result.push(elem) }
      expect(array_string).to eql(result)
end
```

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    it 'is returning an enumerator value when there\'s no blocken given' do
+      result = array_string.my_each
+      expect(result).to be_an Enumerator
+    end
+
+    it 'returns the same array as original mix array' do
+      result = []
+      array_mix.my_each { |elem| result.push(elem) }
+      expect(array_mix).to eql(result)
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: [Optional]
Well, you can test it with the original one. As you have covered the function with the array previously. Something like,
```ruby
it 'returns the same result as the original each does' do
      calculated_result = []
      array_mix.my_each { |elem| calculated_result.push(elem) }
      result = []
      array_mix.each { |elem| result.push(elem) }
      expect(calculated_result).to eql(result)
end
```

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+      my_result = []
+      array_numbers.my_each_with_index { |elem, i| my_result.push(elem + i) }
+      expect(my_result).to eql([10, 3, 5, 7, 5, 10])
+    end
+
+    it 'is returning same array as original string one' do
+      my_result = []
+      array_string.my_each_with_index { |elem, i| my_result.push(elem + i.to_s) }
+      expect(my_result).to eql(%w[100 21 32 43 14 55 arbol6])
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: [Optional]
You can also compare it with the original method by following the previous commented code.

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+    it 'is returning an enumerator value when there\'s no block given' do
+      result = array_string.my_each_with_index
+      expect(result).to be_an Enumerator
+    end
+  end
+
+  describe '#my_select' do
+    it 'returns an enumerator value' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: When it returns an enumerator? I can not get a message from the test.(it should be like `when no block is given, returns an enumerator`).

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+      result = array_string.my_select
+      expect(result).to be_an Enumerator
+    end
+
+    it 'returns a new array with elements equal to 1' do
+      expect(array_numbers.my_select { |elem| elem == 1 }).to eql([1])
+    end
+
+    it 'returns a new array with elements no equal to true' do
+      expect(array_mix.my_select { |elem| elem }).not_to eql([nil, 99])
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: Why it should not contain 99? I tried to execute it with the original select 
```ruby
[nil, true, 99].select {|element| element}   #Returned me [true, 99]
```

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: I just refactored your tests in the format below. Now it seems more readable to me.
```ruby
context 'when block given' do
      it 'returns a new array containing the elements that matches the block' do
        expect(array_numbers.my_select { |elem| elem == 1 }).to eql([1])
      end

      it 'returns a new array that does contain nil/false' do
        expect(array_mix.my_select { |elem| elem }).to eql(array_mix.select { |elem| elem })
      end
end
```
The output seems like.
image

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+      expect(array_string.my_all?(&my_proc_string)).to be false
+    end
+
+    it 'returns true if there are no nil or false elements on given array' do
+      expect(array_numbers.my_all?).to be true
+    end
+
+    it 'returns false if at least there are a nil or false element on given array' do
+      expect(array_mix.my_all?).to be false
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb:90]: I can not see any test case that checks for regexp or class. Please add them.

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      expect(array_string.my_any?(&my_proc_string)).to be true
+    end
+
+    it 'returns true if at least there\'s an element' do
+      expect(array_numbers.my_any?).to be true
+    end
+
+    it 'returns true if there are a nil or false element on given array' do
+      expect(array_mix.my_any?).to be true
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb:128]: Same as above, not test related to regexp or class

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+      expect(array_numbers.my_none?).to be false
+    end
+
+    it 'returns true due to at least there are a nil or false element ' do
+      expect([nil, false].my_none?).to be true
+    end
+
+    it 'returns true due to is an empty array ' do
+      expect([].my_none?).to be true
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb:170]: Nothing related to regexp or class.

File: spec/enumerable_spec.rb [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  end
+
+  describe '#my_count' do
+    it 'returns size array (6) due to there\'s no given block' do
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: It should be generic, for example->
```ruby
it 'returns the number of elements in the array' do
end
```

File: spec/enumerable_spec.rb [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+      expect(array_numbers.my_map).to be_an Enumerator
+    end
+
+    it 'returns same array as original with given block' do
+      expect(array_numbers.my_map { |elem| elem * 10 }).to eq([100, 20, 30, 40, 10, 50])
+    end
+
+    it 'returns same array as original with given proc' do
+      expect(array_numbers.my_map(&my_proc_map)).to eq([100, 20, 30, 40, 10, 50])
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: You can enclose them inside a group using the context shown above.

File: spec/enumerable_spec.rb [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+  end
+
+  describe '#my_inject' do
+    it 'returns all additioned elements plus two' do
+      expect(array_numbers.my_inject(2, :+)).to eq(27)
+    end
+
+    it 'returns all additioned elements in the range plus two' do
+      expect((1..20).my_inject(2, :+)).to eq(212)
+    end
```
Comment (shshamim63) [inline spec/enumerable_spec.rb]: [Optional]
You can compare it using the original inject method.
```ruby
it 'returns the sum of all the elements in the range along with the given initial value' do
      expect((1..20).my_inject(2, :+)).to eq((1..20).inject(2, :+))
end
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Changes Required
___
Hi @bertil291utn :wave: ,
Well done :clap: , you have added almost all the cases, but few are left behind and those are blocking me to approve it now. So let's follow the comments to complete it.
___
### :spades: Mandatory Changes :spades: 
___
1. I can not see anything related to testing in the readme. How to execute the test is not in there. So please add the info.
2. Some of the cases are left behind, mainly with my_all?, my_any?, my_none?. So please add them.
3. Some of your messages are not self-descriptive. I have added some comments to fix them. So please organize the context of your test.
___
### :clubs: Optional Changes :clubs: 
___
1. It is always better to test your methods with the original method. So if any of your methods contains a bug, you can identify it easily.
___
#### Please fix the above-mentioned issue and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback, this bug was solved you said as optional  but anyways it's done

File: spec/enumerable_spec.rb (hunk 2 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback, this bug was solved you said as optional but anyways it's done

File: spec/enumerable_spec.rb (hunk 3 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback, this bug was solved you said as optional but anyways it's done

File: spec/enumerable_spec.rb (hunk 4 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback, this bug was solved

File: spec/enumerable_spec.rb (hunk 5 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Yeah certainly 99 value should be returned and it was changed to 
```
 it 'returns a new array with elements no equal to true' do
      expect(array_mix.my_select { |elem| elem }).not_to eql([nil])
```
Thanks for the feedback, this bug was solved.

File: spec/enumerable_spec.rb (hunk 6 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback, this bug was solved, adding the context block

File: spec/enumerable_spec.rb (hunk 7 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb:90]: Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

File: spec/enumerable_spec.rb (hunk 8 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb:128]: Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

File: spec/enumerable_spec.rb (hunk 9 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb:170]: Thanks for the feedback this bug was solved, adding context/it blocks containing regexp and class check ones

File: spec/enumerable_spec.rb (hunk 10 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback this bug was solved, changing the block message

File: spec/enumerable_spec.rb (hunk 12 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback also this bug was solved, adding original method as compare condition

File: spec/enumerable_spec.rb (hunk 11 above)
Comment (bertil291utn) [reply spec/enumerable_spec.rb]: Thanks for the feedback this bug was solved, adding context block and grouping two `it` blocks

**Review by majovanilla — APPROVED** [review]
Good job implementing the required changes!! :+1: 

Your project looks great and you are good to go and merge!!

### Project APPROVED

You can submit a project completion form, now.

Happy coding 🤓

--Maria Reyes

**Pull request merged**